### ISSUE TITLE
Batch processing

### ISSUE THREAD

**Issue by manthey** [girder/slicer_cli_web#80]
It would be nice to have a standardized way to run a CLI against a batch of files.  Currently, this can be done via a small script calling the appropriate end point for each source file.

This probably will only apply to the first input file/image/item of a CLI.  If a CLI requires multiple input files, and the batch takes a varying file for all inputs, there would need to be someway to ensure the correct correspondence; that is beyond the scope of an initial implementation.

This would have:
- A UI to select a folder instead of a single file/item/image.
- When run, a parent job that would run each job on each image.
- If the parent job is canceled, any sub-job that hasn't been started should never occur.  Those that have started can be allowed to finish (since individual jobs may not have a graceful way to cancel them).
- Jobs may be slow and selected folders could be huge.  If an internal cursor is used to iterate over items, this needs to check that it remains valid for long durations.
- Any output file name that hasn't been explicitly set by the user should use a generated default per input file so that the output files have appropriately distinct names.
- It is desirable to have non-batch jobs run before a large batch job finishes.  This could be done by not creating sub-jobs until needed -- that is, if a sub-job is only created by the parent job when there are no pending jobs (or fewer than some number of pending jobs).
- If multiple batch jobs are run, either the first batch job should finish before subsequent batch jobs, or batch jobs could interleave with each other.  Ideally, if they are interleaved there needs to be some scheduling fairness implemented.
- This needs to work on virtual folders.

Part of the implementation of this could be in the girder worker or jobs plugin as appropriate.

**Comment by sgratzl** [comment]
>     * A UI to select a folder instead of a single file/item/image.

can we assume that the user wants to use the same input parameters besides the file for all the images?

> huge

of which scale are we talking about? 100, 1k, 5k, 10k, 1M?. I don't know whether having a long running DB cursor is a good idea. On the other hand, finding the "next" item in a folder using a query isn't fun either. Moreover, should the list be updated on the fly (like when during the tasks more items are added to the folder) or just statically at the beginning?

>     * Any output file name that hasn't been explicitly set by the user should u
how can a user select an output name. it cannot be only one for all the jobs

>  It is desirable to have non-batch jobs run before a large batch job finishes. This could be done by not creating sub-jobs until needed -- that is, if a sub-job is only created by the parent job when there are no pending jobs (or fewer than some number of pending jobs).

sounds more like priority management of jobs

**Comment by manthey** [comment]
> can we assume that the user wants to use the same input parameters besides the file for all the images?

Yes.

> of which scale are we talking about? 100, 1k, 5k, 10k, 1M?. I don't know whether having a long running DB cursor is a good idea. On the other hand, finding the "next" item in a folder using a query isn't fun either. Moreover, should the list be updated on the fly (like when during the tasks more items are added to the folder) or just statically at the beginning?

I know we want to run this on ~10k items.  

> > ```
> > * Any output file name that hasn't been explicitly set by the user should u
> > ```
> 
> how can a user select an output name. it cannot be only one for all the jobs

Girder lazily attempts to avoid item and folder name duplication by appending a number to a duplicate name.  Further, duplicate names aren't actually prohibited.  You'd end up with a lot of files with poor names like "output.data (<number>)".

We might want (not necessarily in the initial implementation) to allow some sort of template name, e.g., `{name}-output.data`, rather than our auto-generated template name.

**Comment by sgratzl** [comment]
celery has some nice features for creating workflows: https://docs.celeryproject.org/en/latest/userguide/canvas.html e.g. https://docs.celeryproject.org/en/latest/userguide/canvas.html#groups will create a group of tasks that all can be revoked at once and their state observed. 

re task priority: https://docs.celeryproject.org/en/latest/userguide/routing.html#special-routing-options

**Comment by manthey** [comment]
My current thoughts on this:

For the UI, for any single item/image/file field, we currently show an input field and a folder icon to select the item/image/file from the hierarchy widget.  If we had an additional icon to the right for selecting a folder, you could select a folder instead of an item/image/file.  (e.g., 
![image](https://user-images.githubusercontent.com/8781639/100106178-378c8600-2e36-11eb-8acd-185d190071ec.png)).
- If the input record lists file extension(s), then we would filter the contents of the selected folder by the extension list.
- If folders are selected for multiple input item/imagefiles, then we would demand a one-to-one correspondence between the folders (filtered by the extension list and sorted alphabetically by lowerName).
- For output files/directories, allow a jinja-like template syntax for the file name.  Currently, in the slicer_cli_web item view, we populate the output names by default as `{{reference}}-{{task}}-{{parameter}}-{{now}}{{extension}}`, where `reference` is the name of the input referred to by the `reference` attribute (stripped of its extension), `task` is the task name, `parameter` is the output parameter label, `now` is a formatted date string, and `extension` is the first listed file extension.  In the HistomicsUI interface, we use `{{task}}-{{parameter}}{{extension}}`, though `task` is the task title (not name) and `parameter` is the output parameter name  (not label).  I think using the task title and parameter label would be more friendly (and change both slicer_cli_web and HistomicsUI to do this). 
- If any item/image/file inputs are directories, we can create one job to control other jobs.  This job would be cancelable (e.g., we should check if the status changes to canceled and then quit).  Otherwise, it would iterate through the input folder(s) and, if the most recent job it created was not in the inactive state, would create another job with the specific inputs.  By only scheduling another specific job when the last scheduled job is not inactive, this would interleave with non-batch jobs reasonably.
- The batch job should emit status output saying "x of y jobs scheduled".

**Comment by manthey** [comment]
For the UI, see issue #134.

**Comment by manthey** [comment]
The UI  and templating are now separate issues (see #134 and #136).
